alacritty/src/display/window.rs

```diff
@@ -287,47 +262,104 @@ impl Window {
-            let _ = reader.next_frame(&mut buf);
-            Icon::from_rgba(buf, reader.info().width, reader.info().height)
-                .expect("invalid embedded icon format")
+        let activation_token = if let Some(activation_token) = activation_token
+            .map(ActivationToken::from_raw)
+            .or_else(|| event_loop.read_token_from_env())
+        {
+            log::debug!("Activating window with token: {activation_token:?}");
+            // Remove the token from the env.
+            startup_notify::reset_activation_token_env();
+            Some(activation_token)
+        } else {
+            None
         };
 
-        let builder = WinitWindow::default_attributes()
-            .with_name(&identity.class.general, &identity.class.instance)
-            .with_decorations(window_config.decorations != Decorations::None);
+        match event_loop.display_handle().unwrap().as_raw() {
+            #[cfg(all(feature = "x11", not(any(target_os = "macos", windows))))]
+            RawDisplayHandle::Xlib(_) | RawDisplayHandle::Xcb(_) => {
+                Self::get_x11_window(identity, window_config, activation_token, x11_visual)
+            },
+            #[cfg(all(feature = "wayland", not(any(target_os = "macos", windows))))]
+            RawDisplayHandle::Wayland(_) => {
+                Self::get_wayland_window(identity, window_config, activation_token)
+            },
+            _ => unreachable!(),
+        }
+    }
+
+    #[cfg(all(feature = "wayland", not(any(target_os = "macos", windows))))]
+    fn get_wayland_window(
+        identity: &Identity,
+        window_config: &WindowConfig,
+        activation_token: Option<ActivationToken>,
+    ) -> WindowAttributes {
+        let mut attrs = winit::platform::wayland::WindowAttributesWayland::default()
+            .with_name(&identity.class.general, &identity.class.instance);
 
-        #[cfg(feature = "x11")]
-        let builder = builder.with_window_icon(Some(icon));
+        if let Some(activation_token) = activation_token {
+            attrs = attrs.with_activation_token(activation_token);
+        }
 
-        #[cfg(feature = "x11")]
-        let builder = match x11_visual {
-            Some(visual) => builder.with_x11_visual(visual.visual_id() as u32),
-            None => builder,
-        };
+        WindowAttributes::default()
+            .with_decorations(window_config.decorations != Decorations::None)
+            .with_platform_attributes(Box::new(attrs))
+    }
 
-        builder
+    #[cfg(all(feature = "x11", not(any(target_os = "macos", windows))))]
+    fn get_x11_window(
+        identity: &Identity,
+        window_config: &WindowConfig,
+        activation_token: Option<ActivationToken>,
+        x11_visual: Option<X11VisualInfo>,
+    ) -> WindowAttributes {
+        use std::io::Cursor;
+        use winit::icon::RgbaIcon;
+        use winit::platform::x11::WindowAttributesX11;
```

Comment:
Imports should always be at the top, not inside functions. That's the only possible way to keep track of imports, without ending up with duplicates when realizing that something is now needed for all platforms.